File: server/src/collar-drivers/CollarDriver.ts
export interface CollarReading {
  collarId: string
  timestamp: Date
  steps: number          // cumulative since session start
  speedMph: number       // instantaneous
  accelerationG: number  // instantaneous (g-force)
  batteryPct: number
  rawProvider: string    // 'fi' | 'whistle' | 'tractive' | 'ble_generic'
}

export interface CollarDriver {
  readonly provider: string

  /**
   * Authenticate / initialise connection (OAuth token exchange, BLE connect, etc.)
   */
  connect(credentials: Record<string, string>): Promise<void>

  /**
   * Pull the latest reading for a specific collar device.
   */
  fetchReading(externalDeviceId: string): Promise<CollarReading>

  /**
   * Disconnect and clean up resources.
   */
  disconnect(): Promise<void>
}

File: ios/SmartCollar/Collar/CollarDriver.swift
import Foundation

/// Normalized reading emitted by every collar driver
struct CollarReadingBLE {
    let collarId: String
    let timestamp: Date
    let steps: Int
    let speedMph: Double
    let accelerationG: Double
    let batteryPct: Double
}

/// Every collar integration must implement this protocol
protocol CollarDriver: AnyObject {
    var provider: String { get }
    var isConnected: Bool { get }

    /// Called when the user pairs a collar. Implement BLE scan or OAuth initiation here.
    func connect() async throws

    /// Returns the latest reading for the bound device
    func fetchReading() async throws -> CollarReadingBLE

    /// Clean up BLE/network resources
    func disconnect()
}
